crates/loam-cli/src/commands/dev/mod.rs

```diff
@@ -0,0 +1,219 @@
+    Watcher(#[from] notify::Error),
+    #[error(transparent)]
+    Build(#[from] build::Error),
+    #[error("IO error: {0}")]
+    Io(#[from] std::io::Error),
+}
+
+fn canonicalize_path(path: &Path) -> PathBuf {
+    if path.as_os_str().is_empty() {
+        env::current_dir().unwrap_or_else(|_| PathBuf::from("."))
+    } else if path.components().count() == 1 {
+        // Path is a single component, assuming it's a filename
+        env::current_dir()
+            .unwrap_or_else(|_| PathBuf::from("."))
+            .join(path)
+    } else {
+        fs::canonicalize(path).unwrap_or_else(|_| path.to_path_buf())
+    }
+}
+
+#[derive(Clone)]
+struct Watcher {
+    root_env: Arc<PathBuf>,
+    packages: Arc<Vec<PathBuf>>,
+}
+
+impl Watcher {
+    pub fn new(root_env: &Path, packages: &[PathBuf]) -> Self {
+        Self {
+            root_env: Arc::new(canonicalize_path(root_env)),
+            packages: Arc::new(packages.iter().map(|p| canonicalize_path(p)).collect()),
+        }
+    }
+
+    pub fn is_watched(&self, path: &Path) -> bool {
+        let path = canonicalize_path(path);
+        self.packages.iter().any(|p| path.starts_with(p))
+    }
+
+    pub fn is_env_toml(&self, path: &Path) -> bool {
+        canonicalize_path(path) == *self.root_env
+    }
+}
+
+fn is_temporary_file(path: &Path) -> bool {
+    const IGNORED_EXTENSIONS: &[&str] = &["tmp", "swp", "swo"];
+    let file_name = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
+
+    // Vim and vscode temporary files
+    if path
+        .extension()
+        .and_then(|ext| ext.to_str())
+        .map_or(false, |ext| {
+            IGNORED_EXTENSIONS
+                .iter()
+                .any(|&ignored| ext.eq_ignore_ascii_case(ignored))
+        })
+    {
+        return true;
+    }
+
+    // Vim temporary files
+    if file_name.ends_with('~') {
+        return true;
+    }
+
+    // Emacs temporary files
+    if file_name.starts_with('#') && file_name.ends_with('#') {
+        return true;
+    }
+
+    // Add more patterns for other editors as needed
+
+    false
+}
+
+impl Cmd {
+    pub async fn run(&mut self) -> Result<(), Error> {
+        let (tx, mut rx) = mpsc::channel::<Message>(100);
+        let rebuild_state = Arc::new(Mutex::new(false));
+        let workspace_root: &Path = self
+            .build_cmd
+            .manifest_path
+            .parent()
+            .unwrap_or_else(|| Path::new("."));
+        let env_toml_path = workspace_root.join("environments.toml");
+
+        let packages = self
+            .build_cmd
+            .list_packages()?
+            .into_iter()
+            .map(|package| PathBuf::from(package.manifest_path.parent().unwrap().as_str()))
+            .collect::<Vec<_>>();
+
+        let watcher = Watcher::new(&env_toml_path, &packages);
+
+        for package_path in watcher.packages.iter() {
+            eprintln!("Watching {}", package_path.display());
+        }
+
+        let watcher_clone = watcher.clone();
```

Comment:
Why still the initial clone?

Reply:
```error[E0382]: borrow of moved value: `watcher`
    --> crates/loam-cli/src/commands/dev/mod.rs:163:29
     |
127  |         let watcher = Watcher::new(&env_toml_path, &packages);
     |             ------- move occurs because `watcher` has type `dev::Watcher`, which does not implement the `Copy` trait
...
135  |             notify::recommended_watcher(move |res: Result<notify::Event, notify::Error>| {
     |                                         ------------------------------------------------ value moved into closure here
...
147  |                             if watcher.is_watched(path) || watcher.is_env_toml(path) {
     |                                ------- variable moved due to use in closure
...
163  |         for package_path in watcher.packages.iter() {
     |                             ^^^^^^^^^^^^^^^^^^^^^^^ value borrowed here after move
     |
     = note: borrow occurs due to deref coercion to `Vec<PathBuf>`
note: deref defined here
    --> /home/blaine/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/sync.rs:2102:5
     |
2102 |     type Target = T;
     |     ^^^^^^^^^^^
```

Reply:
Yes but why not clone it inside the closure as I have below?

Reply:
Looks like I can avoid cloning if I use the package and env paths directly when setting the notify_watcher to watch them.